lib/vtls/darwinssl.c

```diff
@@ -1996,6 +1996,85 @@ static int verify_cert(const char *cafile, struct Curl_easy *data,
   }
 }
 
+static CURLcode pkp_pin_peer_pubkey(struct SessionHandle *data,
+                                    SSLContextRef ctx,
+                                    const char *pinnedpubkey)
+{  /* Scratch */
+  size_t pubkeylen, realpubkeylen, rsaHeaderLength = 24;
+  unsigned char *pubkey = NULL, *realpubkey = NULL, *rsaHeader = NULL;
+
+  /* Result is returned to caller */
+  CURLcode result = CURLE_SSL_PINNEDPUBKEYNOTMATCH;
+
+  /* if a path wasn't specified, don't pin */
+  if(!pinnedpubkey)
+    return CURLE_OK;
+
+
+  if(!ctx)
+    return result;
+
+  do {
+    SecTrustRef trust;
+    OSStatus ret = SSLCopyPeerTrust(ctx, &trust);
+    if(ret != noErr)
+        break;
+    SecKeyRef keyRef = SecTrustCopyPublicKey(trust);
+
+    CFDataRef publicKeyBits;
+
+    OSStatus success = SecItemExport(keyRef, kSecFormatOpenSSL, 0, NULL,
+                                     &publicKeyBits);
+    if(success != errSecSuccess)
+        break;
+
+    pubkeylen = CFDataGetLength(publicKeyBits);
+    pubkey = CFDataGetBytePtr(publicKeyBits);
+
+    if(pubkeylen == 526) {
+      /* 4096 bit RSA pubkeylen == 526 */
+      const unsigned char rsa4096Header[] = {
+                                       0x30, 0x82, 0x02, 0x22, 0x30, 0x0d,
+                                       0x06, 0x09, 0x2a, 0x86, 0x48, 0x86,
+                                       0xf7, 0x0d, 0x01, 0x01, 0x01, 0x05,
+                                       0x00, 0x03, 0x82, 0x02, 0x0f, 0x00};
+      rsaHeader = rsa4096Header;
+    }
+    else if(pubkeylen == 270) {
+      /* 2048 bit RSA pubkeylen == 270 */
+      const unsigned char rsa2048Header[] = {
+                                       0x30, 0x82, 0x01, 0x22, 0x30, 0x0d,
+                                       0x06, 0x09, 0x2a, 0x86, 0x48, 0x86,
+                                       0xf7, 0x0d, 0x01, 0x01, 0x01, 0x05,
+                                       0x00, 0x03, 0x82, 0x01, 0x0f, 0x00};
+      rsaHeader = rsa2048Header;
+    }
+    else {
+      /* ecDsaSecp256r1Header pubkeylen == 91 is already included?
+       * we assume rest of algorithms do same, so do nothing */
+      result = Curl_pin_peer_pubkey(data, pinnedpubkey, pubkey,
+                                  pubkeylen);
+      break;
+    }
+
+    realpubkeylen = pubkeylen + rsaHeaderLength;
+    realpubkey = malloc(realpubkeylen);
```

Comment:
We can avoid a `malloc` here by stack-allocating a buffer large enough to hold a 4096-bit RSA key. Whether that's a good idea or not is up to @bagder, but if we want to avoid the need to free this pointer it might help.